Update usage of accept_none arg in SQLColumn/TableCheckOperators

denimalpaca opened the pull request:
The accept_none arg is added to the TableCheckOperator and its usage is updated in two operators. Now, the if the arg is True, tests will pass when no rows are returned at either the record or row level. They will return False for relevant checks at the record or row level as well. This means if a record is None and accept_none is True, all tests will pass. Necessary changes to the operators are made to ensure this behavior.

Closes: #30082
Closes: #30081

Commit:
Update usage of accept_none arg in SQLColumn/TableCheckOperators

The accept_none arg is added to the TableCheckOperator and its
usage is updated in two operators. Now, the if the arg is True,
tests will pass when no rows are returned at either the record
or row level. They will return False for relevant checks at the
record or row level as well. This means if a record is None and
accept_none is True, all tests will pass. Necessary changes
to the operators are made to ensure this behavior.

Signed-off-by: Benji Lampel <benjamin@astronomer.io>

Commit:
Update existing tests to match new TableCheckOperator SQL

Signed-off-by: Benji Lampel <benjamin@astronomer.io>

pankajastro reviewed (commented):
I think would be great if we can add tests for this change, wdyt?

Commit:
Add tests for accept_none cases and fix operators wrt tests

Signed-off-by: Benji Lampel <benjamin@astronomer.io>

Commit:
Update SQL in SQLTableCheckOperator

The new sql_check_template SQL was breaking on Postgres
and MySQL tests, so @mag3141592 sent me a new query that
would (hopefully) work for all SQL flavors. Tests also
updated.

Signed-off-by: Benji Lampel <benjamin@astronomer.io>

Commit:
Add subquery alias back to sql_check_template

Signed-off-by: Benji Lampel <benjamin@astronomer.io>

Commit:
Fix type issue in COALESCE.

Signed-off-by: Benji Lampel <benjamin@astronomer.io>

Commit:
More SQL query boolean changes to fix test breaks.

Signed-off-by: Benji Lampel <benjamin@astronomer.io>

Commit:
SQL query changes to COUNT_IF -> COUNT(IFF( to fix test breaks.

Signed-off-by: Benji Lampel <benjamin@astronomer.io>

Commit:
Forgot closing paren

Signed-off-by: Benji Lampel <benjamin@astronomer.io>

Commit:
Maggie rewrote the sql check template for Table operator again, and now it seems good

Signed-off-by: Benji Lampel <benjamin@astronomer.io>

mag3141592 commented on airflow/providers/common/sql/operators/sql.py:
```diff
@@ -572,9 +596,16 @@ class SQLTableCheckOperator(BaseSQLOperator):
+    SELECT
+      '{check_name}' AS check_name,
+      COALESCE(MIN(is_valid), True) AS check_result,
+      COALESCE(MIN(num_rows), False) AS num_subquery_rows
```
```suggestion
      COALESCE(MIN(num_rows), 0) AS num_subquery_rows
```

Commit:
Update airflow/providers/common/sql/operators/sql.py

Co-authored-by: Maggie Stark <maggiestark1@gmail.com>

Commit:
Update tests to reflect Maggie's typo correction.

Signed-off-by: Benji Lampel <benjamin@astronomer.io>

Commit:
Further changes to SQL to try to make it pass the integration tests and updates to tests to simplify strings.

Signed-off-by: Benji Lampel <benjamin@astronomer.io>

Commit:
Fix type issue in SQL check template

Signed-off-by: Benji Lampel <benjamin@astronomer.io>

Commit:
Another SQL template attempt from Maggie.

Signed-off-by: Benji Lampel <benjamin@astronomer.io>

Commit:
Add alias for subquery

Signed-off-by: Benji Lampel <benjamin@astronomer.io>

Commit:
Change COALESCE type to boolean only

Signed-off-by: Benji Lampel <benjamin@astronomer.io>

Commit:
Change MIN to COUNT for postgres

Signed-off-by: Benji Lampel <benjamin@astronomer.io>

Commit:
Merge branch 'main' into sqlcheckoperators-accept-none-argument-updates

Commit:
Merge branch 'main' into sqlcheckoperators-accept-none-argument-updates

Commit:
Merge branch 'main' into sqlcheckoperators-accept-none-argument-updates

Commit:
Merge branch 'main' into sqlcheckoperators-accept-none-argument-updates

Commit:
Merge branch 'main' into sqlcheckoperators-accept-none-argument-updates

Commit:
Merge branch 'main' into sqlcheckoperators-accept-none-argument-updates

Commit:
Merge branch 'main' into sqlcheckoperators-accept-none-argument-updates

pankajastro commented on airflow/providers/common/sql/operators/sql.py:
```diff
@@ -585,41 +622,70 @@ def __init__(
         partition_clause: str | None = None,
         conn_id: str | None = None,
         database: str | None = None,
+        accept_none: bool = True,
```
before these changes, when a query was returning None the task was failing, by setting  `accept_none` default value to true I think will change the old behaviour i.e now a task may not fail. do we need to worry about backward compatibility and maybe set this false by default, wdyt?

denimalpaca replied:
@mag3141592 requested this and may be able to speak better about it, but basically the old behavior is wrong from a correctness standpoint.

mag3141592 replied:
`SQLTableCheckOperator` didn't have an `accept_none` param previously, but `SQLColumnCheckOperator` did and its default is `True`. Ideally these are consistent, but yes that poses an issue with backward compatibility.

So I'm not sure what the priority is, consistency or backward compatibility? I would lean toward consistency, but I also made the request 🤷‍♀️

potiuk replied:
We can treat it as a bug fix. It's really unlikely someone's workflow depends on None being returned here. You'd rather make everything to get a number as result of the check. @eladkal - WDYT?

eladkal replied:
Why can't we deprecate first?
SQLTableCheckOperator - We can add new parameter for with default value False
SQLColumnCheckOperator - Raise deprecation warning that default value will be changed in future versions to False.

In the future we can release common.sql provider with major release that changes the default value.
We are very careful with major releases of common.sql but this one will be fine. WDYT @potiuk ?

potiuk replied:
Sure we can do it this way, but I find it a bit odd to add new parameter and deprecate it immediately. Somehow like - we are adding this new feature there, but BTW it's deprecated. 

If we deprecate it in easy way, there will be no way to distinguish if user deliberately passed False, or it was the default one. 

False is immutable so you cannot use the technique from https://stackoverflow.com/questions/11251119/is-there-any-way-to-know-if-the-value-of-an-argument-is-the-default-vs-user-spe where you could compare identity of passed object. As far as I know. If you make `accept_none=False` the default. for all practical purposes (other than introspecting the source code from the call stack) - `SQLTableCheckOperator(...., accept_none=False)` is indistinguishable from `SQLTableCheckOperator(...., ). Which means that if we deprecate accept_none=False, there will be no way to pass False without a Deprecation warnings - i.e. user who would like to use False will have no way to remove the warning. 

You could potentially pass a sentinel object and treat it as False (and deprecate), but that's really makes things like that complex and ugly (also confusing user):

```
SQLTableCheckOperator(, accept_none: bool | Literal["TREAT_ME_AS_FALSE"] = "TREAT_ME_AS_FALSE") 
```


But if you think it's likely that someone would fall into backwards compatibility there, and it's worth it - yeah. it could be done.

eladkal replied:
> But if you think it's likely that someone would fall into backwards compatibility there, and it's worth it - yeah. it could be done.

We promise to preserve backward compabity unless with major version.
If we think this is a very rare case then we can do a major release without deprecating. Assuming we expect this change to effect very small number of users that can be OK to.

potiuk replied:
Yeah. I am usually in the camp of "breaking change is only determined by how likely it is to change other's workflow" :).  I **think** this case is where it's not very likely somoene treats none as false (though I must say I have never used that operator). 

@pankajastro @denimalpaca - > maybe you have some examples of how people are using the operator and elaborate on how likely it is that the breaking behaviour will kick in ?

pankajastro replied:
sorry, I don't have example I was just worried that this may surprise some users.
@mag3141592 do you have some examples of how we are using this?

mag3141592 replied:
An example of how we're using them, is by testing our incremental table change before loading them into our prod table. If there is no incremental table `SQLTableCheckOperator` currently fails and `SQLColumnCheckOperator` currently passes (because the `accept_none = True` by default). If the defaults are False, then we'd just need to update our end to set `accept_none = True`. 

Sometimes an empty table is indicative of an error and sometimes it isn't, so needed a param for communicating that.

Commit:
Merge branch 'main' into sqlcheckoperators-accept-none-argument-updates
